pkg/dns/dns.go

```diff
@@ -0,0 +1,270 @@
+}
+
+// DNSServer is the implementation of the DNSServerInterface
+type DNSServer struct {
+	bindAddr net.IP
+	port     int
+
+	logger      hclog.Logger
+	client      pbdns.DNSServiceClient
+	connUDP     net.PacketConn
+	listenerTCP net.Listener
+	stopCh      chan (struct{})
+}
+
+// NewDNSServer creates a new DNS proxy server
+func NewDNSServer(p DNSServerParams) (DNSServerInterface, error) {
+	s := &DNSServer{}
+	s.bindAddr = net.ParseIP(p.BindAddr)
+	if s.bindAddr == nil {
+		return nil, fmt.Errorf("error parsing specified dns bind addr '%s'", p.BindAddr)
+	}
+	s.port = p.Port
+	s.client = p.Client
+	s.logger = p.Logger.Named("dns-proxy")
+	return s, nil
+}
+
+// TcpPort is a helper func for the purpose of returning the port
+// that the OS chose if the user specified 0
+func (d *DNSServer) TcpPort() int {
+	if d.listenerTCP == nil {
+		return -1
+	}
+	return int(d.listenerTCP.Addr().(*net.TCPAddr).Port)
+}
+
+// UdpPort is a helper func for the purpose of returning the port
+// that the OS chose if the user specified 0 in the server config
+func (d *DNSServer) UdpPort() int {
+	if d.connUDP == nil {
+		return -1
+	}
+	return int(d.connUDP.LocalAddr().(*net.UDPAddr).Port)
+}
+
+// Run starts the tcp and udp listeners and forwards requests to consul
+func (d *DNSServer) Run() error {
+	if d.port == -1 {
+		return ErrServerDisabled
+	}
+	// 1. Setup udp listener
+	udpAddr := &net.UDPAddr{
+		Port: d.port,
+		IP:   d.bindAddr,
+	}
+	connUDP, err := net.ListenUDP("udp", udpAddr)
+	if err != nil {
+		return fmt.Errorf("error listening for udp: %w", err)
+	}
+	d.connUDP = connUDP
+
+	// 2. Setup tcp listener
+	tcpAddr := &net.TCPAddr{
+		Port: d.port,
+		IP:   d.bindAddr,
+	}
+	listenerTCP, err := net.ListenTCP("tcp", tcpAddr)
+	if err != nil {
+		return fmt.Errorf("error listening to udp: %w", err)
+	}
+	d.listenerTCP = listenerTCP
+
+	// 3. Start go routines to handle dns proxying
+	go d.proxyUDP()
+	go d.proxyTCP()
+
+	// 4. Create stop channel for stopping server
+	d.stopCh = make(chan struct{})
+
+	d.logger.Info("running dns proxy", " udp port", d.UdpPort(), "tcp port", d.TcpPort())
+	return nil
+}
+
+func (d *DNSServer) proxyUDP() {
+	logger := d.logger.Named("udp")
+	for {
+		select {
+		case <-d.stopCh:
+			d.connUDP.Close()
+			return
+		default:
+		}
+		buf := make([]byte, 512)
+		err := d.connUDP.SetReadDeadline(time.Now().Add(time.Second * 10))
+		if err != nil {
+			logger.Error("failure to set read deadline on connection", "error", err)
+			return
+		}
+		bytesRead, addr, err := d.connUDP.ReadFrom(buf)
+		if err != nil {
+			if errors.Is(err, net.ErrClosed) {
+				logger.Info("connection closed")
+				break
+			} else {
+				if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
+					logger.Debug("timeout waiting for read", "error", err)
+				} else {
+					logger.Warn("error reading from conn", "error", err)
+				}
+				continue
+			}
+		}
+		// Parallelize responses
+		go d.queryConsulAndRespondUDP(buf[0:bytesRead], addr)
```

Comment:
How many responses do we predict we'll need to send? Do we need to limit the concurrency here with a worker group or similar?

Reply:
That's a good question. To be honest I think we should be fine since the worst that can happen is that we block on a go routine and the upper bound is just the memory of the system. I am inclined to leave it without one and see how we do. A load test would be a good call here. I'll look into doing that and see how we perform.

Reply:
The intended use is to proxy a single workloads DNS requests and for consul-dataplane to run alongside that workload. The request throughput should be pretty minimal.